src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -1557,6 +1557,520 @@ public static void Main()
+                //         (b1 with { Y = "hello" }).Z.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, @"(b1 with { Y = ""hello"" }).Z").WithLocation(19, 9));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_NullableClone()
+        {
+            var src = @"
+#nullable enable
+
+data class B(string? X)
+{
+    public B? Clone() => new B(X);
+
+    static void M1(B b1)
+    {
+        _ = b1 with { X = null }; // 1
+        (b1 with { X = null }).ToString(); // 2
+    }
+}
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (10,21): warning CS8602: Dereference of a possibly null reference.
+                //         _ = b1 with { X = null }; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "{ X = null }").WithLocation(10, 21),
+                // (11,18): warning CS8602: Dereference of a possibly null reference.
+                //         (b1 with { X = null }).ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "{ X = null }").WithLocation(11, 18));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_MaybeNullClone()
+        {
+            var src = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+data class B(string? X)
+{
+    [return: MaybeNull]
+    public B Clone() => new B(X);
+
+    static void M1(B b1)
+    {
+        _ = b1 with { };
+        _ = b1 with { X = null }; // 1
+        (b1 with { X = null }).ToString(); // 2
+    }
+}
+";
+            var comp = CreateCompilation(new[] { src, MaybeNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (13,21): warning CS8602: Dereference of a possibly null reference.
+                //         _ = b1 with { X = null }; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "{ X = null }").WithLocation(13, 21),
+                // (14,18): warning CS8602: Dereference of a possibly null reference.
+                //         (b1 with { X = null }).ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "{ X = null }").WithLocation(14, 18));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_NotNullClone()
+        {
+            var src = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+data class B(string? X)
+{
+    [return: NotNull]
+    public B? Clone() => new B(X);
+
+    static void M1(B b1)
+    {
+        _ = b1 with { };
+        _ = b1 with { X = null };
+        (b1 with { X = null }).ToString();
+    }
+}
+";
+            var comp = CreateCompilation(new[] { src, NotNullAttributeDefinition });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_NullableClone_NoInitializers()
+        {
+            var src = @"
+#nullable enable
+
+data class B(string? X)
+{
+    public B? Clone() => new B(X);
```

Comment:
It sounds like some upcoming spec changes are probably going to make this necessary to define in metadata if we want to test this scenario.